### PR TITLE
[RzIL] Add concat, append handlers etc.

### PR DESCRIPTION
**Pull request by ogre2007**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

I've just added concat IL opcode handler, and fixed mistake in the definition of the concat bv operation. If this is ok, I'll add another opcodes later

**Test plan**

...

**Closing issues**

linked to #1918

### CODE REVIEW

**Review by wargio — APPROVED** [review]

**Comment by wargio** [comment]
looks like you forgot to change the unit test

**Comment by XVilka** [comment]
Please also rebase.

File: librz/include/rz_il/rzil_opcodes.h [hunk 1]
```diff
@@ -118,6 +118,16 @@ struct rzil_op_cast_t {
 	RzILOp *val; ///< value to cast
 };
 
+
+/**
+ *  \struct rzil_op_concat_t
+ *  \brief op structure for concating bitv
```
Comment (thestr4ng3r) [inline librz/include/rz_il/rzil_opcodes.h]: This does not add any value over the name the op already has, please describe exactly what it does, for example from BAP documentation:
> Appends x and y so that in the resulting vector the vector x occupies the most significant part and y the least significant part.

**Review by thestr4ng3r — CHANGES_REQUESTED** [review]
It should be called "append" to correspond to BAP: http://binaryanalysisplatform.github.io/bap/api/master/bap-core-theory/Bap_core_theory/Theory/module-type-Minimal/index.html#val-append

File: librz/include/rz_il/rzil_opcodes.h (hunk 1 above)
Comment (XVilka) [reply librz/include/rz_il/rzil_opcodes.h]: Yes, append operates on two bitvectors, while concat operates on the list of bitvectors.

**Review by wargio — CHANGES_REQUESTED** [review]
unit test needs to be changed (https://github.com/rizinorg/rizin/blob/dev/test/unit/test_il_definitions.c) and also you need to add `void *rz_il_handler_concat(RzILVM *vm, RzILOp *op, RzILOpArgType *type);` definition in vm_layer.c

**Comment by ogre2007** [comment]
I've renamed concat into append, and added new concat operation, but there is not RzList ARG type in opcodes, so its not correct currently. Should I add new ARG_TYPE?

File: librz/include/rz_il/rzil_vm.h [hunk 2]
```diff
@@ -100,6 +100,54 @@ RZ_API RzILEffect *rz_il_evaluate_effect(RzILVM *vm, RzILOp *op, RzILOpArgType *
...
+void *rz_il_handler_jmp(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_goto(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_seq(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_branch(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+
+void *rz_il_handler_load(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_store(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+
+//TODO: remove me when all the handlers are implemented
+void *rz_il_handler_unimplemented(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
```
Comment (wargio) [inline librz/include/rz_il/rzil_vm.h]: i have removed them for a reason. just add the concat/append at the beginning of vm_layer.c

File: librz/il/definitions/bitvector.c [hunk 3]
```diff
@@ -251,24 +251,48 @@ RZ_API RZ_OWN RzILBitVector *rz_il_bv_cut_tail(RZ_NONNULL RzILBitVector *bv, ut3
...
+			tmp = ret;
+			continue;
+		}
+		ret = rz_il_bv_append(ret, bv);
+		rz_il_bv_free(tmp);
+		tmp = ret;
+	}
+	tmp = NULL;
+	return ret;
+}
```
Comment (wargio) [inline librz/il/definitions/bitvector.c]: i'm not sure about this. @XVilka

**Review by wargio — CHANGES_REQUESTED** [review]

File: librz/include/rz_il/rzil_vm.h (hunk 2 above)
Comment (ogre2007) [reply librz/include/rz_il/rzil_vm.h]: Sure! But the question is how to provide RzList as argument for concat inside the operation object, bcs there is no argtype for it

File: librz/il/definitions/bitvector.c (hunk 3 above)
Comment (ogre2007) [reply librz/il/definitions/bitvector.c]: It actually works though. I can do it w/o append function, but it takes two foreach loops: 1st for collecting actual lengths, then we can allocate resulting bv, then we copy bvs content in the second foreach loop.

File: librz/include/rz_il/rzil_vm.h (hunk 2 above)
Comment (wargio) [reply librz/include/rz_il/rzil_vm.h]: i think you need to edit the RzILOpConcat structure to have an RzList, but i would wait for @XVilka for his input. i'm not sure if that is the right way to implement this.

File: librz/il/definitions/bitvector.c (hunk 3 above)
Comment (wargio) [reply librz/il/definitions/bitvector.c]: looks correct, but the problem is if this is actually needed, or not. so this is why i asked to wait for him

**Comment by XVilka** [comment]
On the one hand, such an operation could be emulated by calling multiple `append` operations. On the other hand, it might become handy for uplifting more complex instructions. I guess we can add a new `list` type for the arguments. @thestr4ng3r your thoughts?

**Comment by thestr4ng3r** [comment]
From what I can tell, `RzILOpArgType` is only needed when an op **returns** a list, but not if it takes one. Or are there ops in BAP that can return lists of bitvectors? That would seem overcomplicated to me.
Regardless, I would just omit the concat completely for now. `RzList` is a linked list anyway and chaining multiple append ops would actually be one too.

File: librz/il/definitions/bitvector.c [hunk 4]
```diff
@@ -251,12 +251,12 @@ RZ_API RZ_OWN RzILBitVector *rz_il_bv_cut_tail(RZ_NONNULL RzILBitVector *bv, ut3
 }
 
 /**
- * Concat bv1 and bv2 to get new bitvector
```
Comment (XVilka) [inline librz/il/definitions/bitvector.c:264]: Append then

File: test/unit/test_il_definitions.c [hunk 5]
```diff
@@ -179,6 +179,7 @@ bool test_rzil_bv_algorithm(void) {
 	RzILBitVector *x, *y;
 	RzILBitVector *result;
 	RzILBitVector *add, *sub, *mul, *div, *mod;
+
```
Comment (XVilka) [inline test/unit/test_il_definitions.c]: Unnecessary change

**Review by XVilka — CHANGES_REQUESTED** [review]

**Comment by XVilka** [comment]
@ogre2007 please rebase and address my comments, we can merge it then.

**Comment by ogre2007** [comment]
ok, Ill do it today

**Review by XVilka — APPROVED** [review]

File: librz/il/vm_layer.c [hunk 6]
```diff
@@ -177,8 +177,8 @@ RZ_API bool rz_il_vm_init(RzILVM *vm, ut64 start_addr, ut32 addr_size, ut32 data
 	vm->op_handler_table[RZIL_OP_SLE] = &rz_il_handler_unimplemented; // &rz_il_handler_sle;
 	vm->op_handler_table[RZIL_OP_ULE] = &rz_il_handler_unimplemented; // &rz_il_handler_ule;
 	vm->op_handler_table[RZIL_OP_CAST] = &rz_il_handler_cast;
-	vm->op_handler_table[RZIL_OP_CONCAT] = &rz_il_handler_unimplemented; // &rz_il_handler_concat;
-	vm->op_handler_table[RZIL_OP_APPEND] = &rz_il_handler_unimplemented; // &rz_il_handler_append;
+	vm->op_handler_table[RZIL_OP_CONCAT] = &rz_il_handler_unimplemented; // &rz_il_handler_concat
+	vm->op_handler_table[RZIL_OP_APPEND] = &rz_il_handler_append; // &rz_il_handler_append;
```
Comment (wargio) [inline librz/il/vm_layer.c]:
```suggestion
	vm->op_handler_table[RZIL_OP_APPEND] = &rz_il_handler_append;
```

File: librz/include/rz_il/rzil_vm.h [hunk 7]
```diff
@@ -100,6 +100,52 @@ RZ_API RzILEffect *rz_il_evaluate_effect(RzILVM *vm, RzILOp *op, RzILOpArgType *
...
+
```
Comment (wargio) [inline librz/include/rz_il/rzil_vm.h]:
```suggestion
```

Comment (wargio) [reply librz/include/rz_il/rzil_vm.h]: please add the definition (`void *rz_il_handler_append(RzILVM *vm, RzILOp *op, RzILOpArgType *type);`) in vm_layer.c

This header mush **NOT** have any of these definitions there.

**Review by wargio — CHANGES_REQUESTED** [review]

**Review by wargio — APPROVED** [review]
